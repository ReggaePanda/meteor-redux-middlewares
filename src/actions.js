import { createAction } from './utils';

export const REGISTER_REACTIVE_SOURCE = 'REGISTER_REACTIVE_SOURCE';
export const registerReactiveSource = createAction(REGISTER_REACTIVE_SOURCE);

export const STOP_SUBSCRIPTION = 'STOP_SUBSCRIPTION';
export const stopSubscription = createAction(STOP_SUBSCRIPTION);

export const START_SUBSCRIPTION = 'START_SUBSCRIPTION';
export const startSubscription = createAction(START_SUBSCRIPTION);
