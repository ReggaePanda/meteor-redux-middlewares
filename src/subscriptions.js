import runContextual from './runContextual';

import {
  STOP_SUBSCRIPTION,
  START_SUBSCRIPTION,
  stopSubscription,
} from './actions';

import {
  actionCase,
  errorWith,
  hasGet,
  hasKey,
  hasSubscribe,
  stringPayload,
} from './utils';

const subscriptions = {};
const computations = {};

export default tracker => ({ dispatch }) => next => action => {
  const throwIfNot = errorWith(action);

  if (action.type === STOP_SUBSCRIPTION) {
    const stop = () => {
      throwIfNot(stringPayload,
        'A stopSubscription action needs a string payload to identify a subscription'
      );

      if (subscriptions[action.payload]) {
        const subscriptionId = subscriptions[action.payload].subscriptionId;

        computations[subscriptionId].stop();
        subscriptions[action.payload].stop();
      }
    };

    runContextual(stop);
  }

  if (action.type === START_SUBSCRIPTION) {
    const start = () => {
      throwIfNot(hasSubscribe,
        'A startSubscription action needs a `subscribe` function to start a subscription'
      );

      throwIfNot(hasKey,
        'A startSubscription action needs a `key` string to identify a subscription'
      );

      throwIfNot(hasGet,
        'A startSubscription action needs a `get` function to load data'
      );

      const { key, subscribe } = action.payload;

      if (subscriptions[key]) {
        dispatch(stopSubscription(key));
      }

      const subscription = subscribe();
      const { subscriptionId } = subscription;

      subscriptions[key] = subscription;
      computations[subscriptionId] = tracker.autorun(() => {
        const ready = subscription.ready();

        if (ready) {
          dispatch({
            type: `${actionCase(key)}_SUBSCRIPTION_CHANGED`,
            payload: action.payload.get(),
          });
        }

        dispatch({
          type: `${actionCase(key)}_SUBSCRIPTION_READY`,
          payload: {
            ready,
            data: action.payload.onReadyData
              ? action.payload.onReadyData()
              : {},
          },
        });
      });
    };

    runContextual(start);
  }

  return next(action);
};
