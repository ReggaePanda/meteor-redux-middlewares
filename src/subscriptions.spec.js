import configureMockStore from 'redux-mock-store';
import subscriptions from './subscriptions';
import { stopSubscription, startSubscription } from './actions';

/**
 * See src/__mocks__/mocks.js
 *
 * The results array imported here is used to hold tags pushed from various
 * functions to ensure the sequence is as expected
 * (the tags are in pirate speak)
 *
 */
const {
  get,
  results,
  subscribe,
  tracker,
} = require('mocks')();

const onReadyData = 'im ready, but me eyepatch is lost';
const subKey = 'mates';

const stop = stopSubscription(subKey);
const start = startSubscription({
  onReadyData: () => {
    results.push(onReadyData);
    return onReadyData;
  },
  key: subKey,
  subscribe,
  get,
});

const store = configureMockStore([subscriptions(tracker)])({});
store.dispatch(start);
store.dispatch(start);
store.dispatch(stop);

it('should handle a startSubscription action', () => {
  expect(store.getActions()).toMatchSnapshot();
});

it('should fire in the correct sequence', () => {
  expect(results).toMatchSnapshot();
});
