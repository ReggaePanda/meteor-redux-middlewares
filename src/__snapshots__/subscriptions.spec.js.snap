exports[`test should fire in the correct sequence 1`] = `
Array [
  "yarg, i subscribed!",
  "i got me data!",
  "im ready, but me eyepatch is lost",
  "my computation be stopped now!",
  "my subscription be stopped now!",
  "yarg, i subscribed!",
  "i got me data!",
  "im ready, but me eyepatch is lost",
  "my computation be stopped now!",
  "my subscription be stopped now!",
]
`;

exports[`test should handle a startSubscription action 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "id": "treasure",
      },
    ],
    "type": "MATES_SUBSCRIPTION_CHANGED",
  },
  Object {
    "payload": Object {
      "data": "im ready, but me eyepatch is lost",
      "ready": true,
    },
    "type": "MATES_SUBSCRIPTION_READY",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "get": [Function],
      "key": "mates",
      "onReadyData": [Function],
      "subscribe": [Function],
    },
    "type": "START_SUBSCRIPTION",
  },
  Object {
    "meta": Object {},
    "payload": "mates",
    "type": "STOP_SUBSCRIPTION",
  },
  Object {
    "payload": Array [
      Object {
        "id": "treasure",
      },
    ],
    "type": "MATES_SUBSCRIPTION_CHANGED",
  },
  Object {
    "payload": Object {
      "data": "im ready, but me eyepatch is lost",
      "ready": true,
    },
    "type": "MATES_SUBSCRIPTION_READY",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "get": [Function],
      "key": "mates",
      "onReadyData": [Function],
      "subscribe": [Function],
    },
    "type": "START_SUBSCRIPTION",
  },
  Object {
    "meta": Object {},
    "payload": "mates",
    "type": "STOP_SUBSCRIPTION",
  },
]
`;
